controllers/jujucluster_controller.go

```diff
@@ -687,64 +702,122 @@ func (r *JujuClusterReconciler) getBootstrapJob(ctx context.Context, jujuCluster
 }
 
 func createCloud(ctx context.Context, jujuCluster *infrastructurev1beta1.JujuCluster, client *juju.Client) error {
-	vsphereRegion := cloud.Region{
-		Name:     "Boston",
-		Endpoint: jujuCluster.Spec.CloudEndpoint,
+	log := log.FromContext(ctx)
+	// Convert spec.Cloud to a juju cloud.Cloud by marshalling to JSON, then unmarshalling to juju cloud type
+	// This is done because juju cloud struct contains interface{} for some types, which is not allowed in kube-builder structs
+	// so a new cloud struct had to be provided to conform with kube-builder expectations
+	cloudJson, err := json.Marshal(jujuCluster.Spec.Cloud)
+	if err != nil {
+		log.Error(err, "error marshalling cloud")
+		return err
 	}
+	log.Info("successfully marshalled cloud")
 
-	vsphereCloud := cloud.Cloud{
-		Name:      jujuCluster.Name,
-		Type:      "vsphere",
-		Endpoint:  jujuCluster.Spec.CloudEndpoint,
-		Regions:   []cloud.Region{vsphereRegion},
-		AuthTypes: []cloud.AuthType{"userpass"},
+	jujuCloud := &cloud.Cloud{}
+	err = json.Unmarshal(cloudJson, jujuCloud)
+	if err != nil {
+		log.Error(err, "error unmarshalling cloud")
+		return err
 	}
+	log.Info("successfully unmarshalled cloud")
 
 	// Force must be true when adding a non-k8s cloud to a in-k8s juju controller
 	addCloudInput := juju.AddCloudInput{
-		Cloud: vsphereCloud,
+		Cloud: *jujuCloud,
 		Force: true,
 	}
 	return client.Clouds.AddCloud(ctx, addCloudInput)
 }
 
-func createCredential(ctx context.Context, cloudSecret *kcore.Secret, jujuCluster *infrastructurev1beta1.JujuCluster, client *juju.Client) error {
-	username := string(cloudSecret.Data["username"][:])
-	password := string(cloudSecret.Data["password"][:])
-	vmfolder := string(cloudSecret.Data["vmfolder"][:])
-	credential := cloud.NewCredential("userpass", map[string]string{
-		"user":     username,
-		"password": password,
-		"vmfolder": vmfolder,
-	})
-	addCredInput := juju.AddCredentialInput{
-		Credential:     credential,
-		CredentialName: jujuCluster.Name,
-		CloudName:      jujuCluster.Name,
-	}
-	return client.Credentials.AddCredential(ctx, addCredInput)
+func createCredential(ctx context.Context, credentialSecret *kcore.Secret, jujuCluster *infrastructurev1beta1.JujuCluster, client *juju.Client) error {
+	log := log.FromContext(ctx)
+	// capi secrets (kubeconfig, etc) generally have a value key containing the data, so that format is also used for the credential secret
+	data, ok := credentialSecret.Data["value"]
+	if !ok {
+		log.Error(nil, "credential secret does not have correct format, missing key", "key", "value")
+		return errors.New("invalid format")
+	}
+
```

Comment:
Leverage built-in juju functionality to parse the credentials file contained in the secret data